design/build-cache.md

```diff
@@ -0,0 +1,201 @@
+ss# Build cache
+
+## Context
+
+Users require a way to cache resources between jobs within a single pipeline or consecutive executions of a particular pipeline in order to speed up build time.
+
+## Status
+
+- 29 June 2018: Proposal submitted
+
+## Proposal
+
+Provide a mechanism for transparent caching of resources between jobs within a single pipeline, or between consecutive executions of a particular pipeline.
+
+1. Content to be cached can be specified by the user.
+2. Caching is scoped. 
+   1. Job level caching
+   2. Event level caching
+   3. Pipeline level caching (static caching, in a sense)
+3. Well used caching patterns can be offered as predefined options.
+   1. Dependency caching for build systems such as npm, sbt and cargo.
+
+The build cache and cache usage is configurable via Yaml and backed by an internal sd-command and caching service.
+
+## Details
+
+### Yaml specification
+
+```yml
+shared:
+  # Caches defined in the shared section can be pipeline or exection scoped.
+  cache:
+    # A pipeline scoped cache.
+    pipeline:  
+      # The caching targets
+      target: ["build/**", "~/.sbt"]
+    # An event scoped cache
+    event:
+      target: ["target/deployment.jar"]
+    # A job scoped cache 
+    my-cache-1:
```

Comment:
To adjust to other formats, how about below format?
```
job:
  my-cache-1:
    target: ["build/some-artifact.zip"]
```

On the other hand, `target` key looks verbose. I think these can be more simply.

```
  cache:
    # A pipeline scoped cache.
    pipeline: ["build/**", "~/.sbt"]
    # An event scoped cache
    event: ["target/deployment.jar"]
    # A job scoped cache 
    job:
      my-cache-1: ["build/some-artifact.zip"]
      my-cache-2: ["build/some-artifact2.zip"]
```

Reply:
Thanks @tk3fftk, that's a great suggestion. Will reflect change the syntax to reflect this.

- [ ] @elyzion Update job level cache definition example syntax.